File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n=nums.size();
        int pivot=-1;
        for(int i=n-1;i>=1;i--){
            if(nums[i]>nums[i-1]){
                pivot=i-1;
                break;
            }
        }
        if(pivot==-1) reverse(nums.begin(),nums.end());
        else{
            int first_bigger=-1;
            for(int i=n-1;i>pivot;i--){
                if(nums[i]>nums[pivot]){
                    first_bigger=i;
                    break;
                }
            }
            swap(nums[pivot],nums[first_bigger]);
            reverse(nums.begin()+pivot+1,nums.end());
        }
    }
};

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int i=nums.length-2,pivot=-1;
        while(i>=0){
            if(nums[i]<nums[i+1]){
                break;
            }
            i--;
        }
        // System.out.println(i);
        if(i==-1){
            reverse(nums,i+1,nums.length-1);
            // System.out.println("YES");
        }
        else{
        int j=nums.length-1;
        while(j>i){
            if(nums[j]>nums[i]){
                break;
            }
            j--;
        }
        // System.out.println(j);
            int temp=nums[i];
            nums[i]=nums[j];
            nums[j]=temp;
        reverse(nums,i+1,nums.length-1);
        }
                
    }
    void reverse(int[] nums, int start ,int end ){
        while(start<end){
            int temp=nums[start];
            nums[start]=nums[end];
            nums[end]=temp;
            start++;
            end--;
        }
    }
}